### PR TITLE
Improve keystroke tests

### PR DESCRIPTION
**Pull request by VojtaStavik**
#### In this PR:
- The tests for keystroke events now run synchronously

#no_changelog

### CODE REVIEW

File: Sources/Client/Client/Client.swift [hunk 1]
```diff
@@ -133,6 +133,11 @@ public final class Client: Uploader {
         self.onUserUpdateObservers.values.forEach({ $0(newUser) })
     }
     
+    #if DEBUG
+    /// Called when a new outgoing event is about to be sent. Meant to be used only for testing purposes.
+    var outgoingEventsTestLogger: ((EventType) -> Void)?
+    #endif
```
Comment (buh) [inline Sources/Client/Client/Client.swift:139]: This smells bad and looks like a hardcoded solution, but it's really nice for test usage.

File: Sources/Client/Model/Channel/Channel+Requests.swift [hunk 2]
```diff
@@ -145,24 +151,27 @@ public extension Channel {
     
     @discardableResult
     internal func keystroke(client: Client,
-                            timerType: Timer.Type,
+                            timerType: Timer.Type = DefaultTimer.self,
```
Comment (buh) [inline Sources/Client/Model/Channel/Channel+Requests.swift:154]: This is the internal method and it's only for tests. In the same public method explicitly is using `DefaultTimer.self`, so it's useless here.

File: Sources/Client/Model/Channel/Channel.swift [hunk 3]
```diff
@@ -128,6 +128,9 @@ public final class Channel: Codable {
     /// Checks for the channel data encoding is empty.
     var isEmpty: Bool { extraData == nil && members.isEmpty && invitedMembers.isEmpty }
     
+    /// Returns the current timestamp. Can be replaced in tests with mock time, if needed.
+    var currentTime: () -> Date = { Date() }
```
Comment (buh) [inline Sources/Client/Model/Channel/Channel.swift:132]: I like this very much. I thought about how to user `VirtualTime` but didn't find a good solution.

File: Sources/Client/Model/Channel/Channel+Requests.swift [hunk 4]
```diff
@@ -12,6 +12,12 @@ import Foundation
 
 public extension Channel {
     
+    /// The number of seconds from the last `typingStart` event until the `typingStop` event is automatically sent.
+    static var startTypingEventTimeout: TimeInterval = 15
```
Comment (buh) [inline Sources/Client/Model/Channel/Channel+Requests.swift:16]: I think we can change it to 5.

File: Tests/Client/Unit Tests/Channel+KeystrokeTests.swift [hunk 5]
```diff
@@ -11,107 +11,46 @@ import XCTest
...
+        // 4. User stopped typing, the `typingStop` event should be sent
+        time.run(numberOfSeconds: 6)
+        XCTAssertEqual(eventLogger, [.typingStart, .typingStart, .typingStop])
     }
     
     private func sendKeystroke(in channel: Channel) {
         channel.keystroke(client: client, timerType: VirtualTimeTimer.self) { _ in }
     }
-    
-    func test_channel_stopTyping() throws {
```
Comment (buh) [inline Tests/Client/Unit Tests/Channel+KeystrokeTests.swift:81]: Why did you remove it?

File: Sources/Client/Client/Client.swift (hunk 1 above)
Comment (VojtaStavik) [reply Sources/Client/Client/Client.swift:139]: You must play with the cards you have 🤷‍♂️ 😄

File: Tests/Client/Unit Tests/Channel+KeystrokeTests.swift (hunk 5 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Channel+KeystrokeTests.swift:81]: Because it's already cover the test above

File: Sources/Client/Model/Channel/Channel+Requests.swift (hunk 2 above)
Comment (VojtaStavik) [reply Sources/Client/Model/Channel/Channel+Requests.swift:154]: Useless is a strong word. Unnecessary? Yes :)

File: Sources/Client/Model/Channel/Channel+Requests.swift (hunk 4 above)
Comment (VojtaStavik) [reply Sources/Client/Model/Channel/Channel+Requests.swift]: fixed in f0cf4f3

File: Sources/Client/Model/Channel/Channel+Requests.swift (hunk 2 above)
Comment (buh) [reply Sources/Client/Model/Channel/Channel+Requests.swift:154]: I'm sorry. Yes, unnecessary!

**Comment by buh** [comment]
@VojtaStavik you can merge it too
